File: Editor/Resources/GetSelectionResource.cs
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;

namespace McpUnity.Resources
{
    /// <summary>
    /// Resource for getting the current editor selection state.
    /// Returns the active GameObject, all selected GameObjects, and selected assets.
    /// </summary>
    public class GetSelectionResource : McpResourceBase
    {
        public GetSelectionResource()
        {
            Name = "get_selection";
            Description = "Retrieves the current Unity Editor selection (active object, selected GameObjects, selected assets)";
            Uri = "unity://selection";
        }

        public override JObject Fetch(JObject parameters)
        {
            JObject activeGameObject = null;
            GameObject activeGo = Selection.activeGameObject;
            if (activeGo != null)
            {
                activeGameObject = new JObject
                {
                    ["name"] = activeGo.name,
                    ["instanceId"] = activeGo.GetInstanceID(),
                    ["path"] = GameObjectToolUtils.GetGameObjectPath(activeGo)
                };
            }

            // All selected GameObjects (scene objects)
            JArray selectedGameObjects = new JArray();
            foreach (GameObject go in Selection.gameObjects)
            {
                selectedGameObjects.Add(new JObject
                {
                    ["name"] = go.name,
                    ["instanceId"] = go.GetInstanceID(),
                    ["path"] = GameObjectToolUtils.GetGameObjectPath(go)
                });
            }

            // Selected assets in the Project window
            JArray selectedAssets = new JArray();
            string[] guids = Selection.assetGUIDs;
            foreach (string guid in guids)
            {
                string assetPath = AssetDatabase.GUIDToAssetPath(guid);
                if (string.IsNullOrEmpty(assetPath))
                    continue;

                Object asset = AssetDatabase.LoadMainAssetAtPath(assetPath);
                selectedAssets.Add(new JObject
                {
                    ["name"] = asset != null ? asset.name : System.IO.Path.GetFileNameWithoutExtension(assetPath),
                    ["path"] = assetPath,
                    ["guid"] = guid,
                    ["type"] = asset != null ? asset.GetType().Name : "Unknown"
                });
            }

            return new JObject
            {
                ["success"] = true,
                ["message"] = $"Selection: {selectedGameObjects.Count} GameObject(s), {selectedAssets.Count} asset(s)",
                ["activeGameObject"] = activeGameObject != null ? activeGameObject : JValue.CreateNull(),
                ["selectedGameObjects"] = selectedGameObjects,
                ["selectedAssets"] = selectedAssets
            };
        }
    }
}

File: Server~/src/resources/getSelectionResource.ts
import { McpUnity } from '../unity/mcpUnity.js';
import { Logger } from '../utils/logger.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { ReadResourceResult } from '@modelcontextprotocol/sdk/types.js';

const resourceName = 'get_selection';
const resourceUri = 'unity://selection';
const resourceMimeType = 'application/json';

/**
 * Registers the Selection resource with the MCP server.
 * Returns the current Unity Editor selection: active GameObject,
 * all selected GameObjects, and selected assets in the Project window.
 */
export function registerGetSelectionResource(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering resource: ${resourceName}`);

  server.resource(
    resourceName,
    resourceUri,
    {
      description: 'Retrieve the current Unity Editor selection (active object, selected GameObjects, selected assets)',
      mimeType: resourceMimeType
    },
    async () => {
      try {
        return await resourceHandler(mcpUnity);
      } catch (error) {
        logger.error(`Error handling resource ${resourceName}: ${error}`);
        throw error;
      }
    }
  );
}

async function resourceHandler(mcpUnity: McpUnity): Promise<ReadResourceResult> {
  const response = await mcpUnity.sendRequest({
    method: resourceName,
    params: {}
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.RESOURCE_FETCH,
      response.message || 'Failed to fetch selection from Unity Editor'
    );
  }

  const selectionData = {
    activeGameObject: response.activeGameObject ?? null,
    selectedGameObjects: response.selectedGameObjects ?? [],
    selectedAssets: response.selectedAssets ?? []
  };

  return {
    contents: [
      {
        uri: resourceUri,
        mimeType: resourceMimeType,
        text: JSON.stringify(selectionData, null, 2)
      }
    ]
  };
}
